File: Basic/Chapter3/SPFA.py
"""
SPFA
queue optimization of Bellman-Ford algorithm
"""
import collections
from math import inf

"""
851. spfa求最短路
https://www.acwing.com/problem/content/853/
"""

n, m = map(int, input().split())
g = [[] for _ in range(n + 1)]
dist = [inf] * (n + 1)
state = [False] * (n + 1)
for _ in range(m):
    x, y, z = map(int, input().split())
    g[x].append((y, z))


def spfa_shortest_path():
    global n
    dq = collections.deque()

    dq.append(1)
    dist[1] = 0
    state[1] = True

    while dq:
        x = dq.pop()
        state[x] = False
        for y, z in g[x]:
            if dist[y] > dist[x] + z:
                dist[y] = dist[x] + z
                if not state[y]:
                    # don't repeat add
                    state[y] = True
                    dq.append(y)

    return dist[n]


res = spfa_shortest_path()
print(res if res != inf else "impossible")

File: Basic/Chapter3/SPFA.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type node struct {
	y, z int
}

var (
	n, m int
	g    [][]node
)

func spfa() (res int) {
	dq := make([]int, 0)
	dist := make([]int, n+1)
	state := make([]bool, n+1)
	dq = append(dq, 1)
	dist[1] = 0
	for i := 2; i <= n; i++ {
		dist[i] = 0x3f3f3f3f
	}
	state[1] = true

	for len(dq) > 0 {
		x := dq[0]
		dq = dq[1:]
		state[x] = false
		for _, nd := range g[x] {
			y, z := nd.y, nd.z
			if dist[y] > dist[x]+z {
				dist[y] = dist[x] + z
				if !state[y] {
					state[y] = true
					dq = append(dq, y)
				}
			}
		}
	}

	return dist[n]
}

func main() {
	in := bufio.NewReader(os.Stdin)
	fmt.Fscan(in, &n, &m)
	g = make([][]node, n+1)
	//for i := 0; i < n; i++ {
	//	g[i] = make([]node, 0)
	//}
	for i := 0; i < m; i++ {
		var x, y, z int
		fmt.Fscan(in, &x, &y, &z)
		g[x] = append(g[x], node{y, z})
	}

	res := spfa()
	if res == 0x3f3f3f3f {
		fmt.Println("impossible")
	} else {
		fmt.Println(res)
	}
}
